src/controller/CHIPDeviceController.cpp

```diff
@@ -1209,192 +1143,167 @@ CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(const ByteSpan & attestat
-    ReturnErrorOnFailure(dacVerifier->VerifyNodeOperationalCSRInformation(NOCSRElements, attestationChallenge, AttestationSignature,
-                                                                          dacPubkey, device->GetCSRNonce()));
-
-    mOperationalCredentialsDelegate->SetNodeIdForNextNOCRequest(device->GetDeviceId());
-
-    FabricInfo * fabric = mSystemState->Fabrics()->FindFabricWithIndex(mFabricIndex);
-    mOperationalCredentialsDelegate->SetFabricIdForNextNOCRequest(fabric->GetFabricId());
+    OperationalCredentials::Commands::AddTrustedRootCertificate::Type request;
+    request.rootCertificate = rcac;
+    return SendCommand<OperationalCredentialsCluster>(device, request, OnAddTrustedRootCertificate);
+}
 
-    return mOperationalCredentialsDelegate->GenerateNOCChain(NOCSRElements, AttestationSignature, device->GetDAC(), ByteSpan(),
-                                                             ByteSpan(), &mDeviceNOCChainCallback);
+TRAMPOLINE(AddTrustedRootCertificate, const app::DataModel::NullObjectType &);
+void DeviceCommissioner::OnAddTrustedRootCertificate(const app::DataModel::NullObjectType & data)
+{
+    StopSessionIf(mState != State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    StopSessionIf(mDeviceBeingCommissioned == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    StopSessionIf(SendOperationalCertificate(mDeviceBeingCommissioned, mDeviceBeingCommissioned->GetNOCCert(),
+                                             mDeviceBeingCommissioned->GetICACert()));
 }
 
 CHIP_ERROR DeviceCommissioner::SendOperationalCertificate(CommissioneeDeviceProxy * device, const ByteSpan & nocCertBuf,
                                                           const ByteSpan & icaCertBuf)
 {
-    VerifyOrReturnError(device != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
-    chip::Controller::OperationalCredentialsCluster cluster;
-    cluster.Associate(device, 0);
+    OperationalCredentials::Commands::AddNOC::Type request;
+    request.NOCValue      = mDeviceBeingCommissioned->GetNOCCert();
+    request.ICACValue     = MakeOptional(mDeviceBeingCommissioned->GetICACert());
+    request.IPKValue      = ByteSpan(nullptr, 0);
+    request.caseAdminNode = mLocalId.GetNodeId();
+    request.adminVendorId = mVendorId;
 
-    Callback::Cancelable * successCallback = mNOCResponseCallback.Cancel();
-    Callback::Cancelable * failureCallback = mOnCertFailureCallback.Cancel();
+    return SendCommand<OperationalCredentialsCluster>(mDeviceBeingCommissioned, request, OnAddNOCResponse);
+}
 
-    ReturnErrorOnFailure(cluster.AddNOC(successCallback, failureCallback, nocCertBuf, icaCertBuf, ByteSpan(nullptr, 0),
-                                        mLocalId.GetNodeId(), mVendorId));
+TRAMPOLINE(AddNOCResponse, const OperationalCredentials::Commands::NOCResponse::DecodableType &);
+void DeviceCommissioner::OnAddNOCResponse(const OperationalCredentials::Commands::NOCResponse::DecodableType & data)
```

Comment:
This also lost its logging....  Several levels of it, in fact.

Reply:
I have added it back.